Migrate several indexing and compaction integration tests to embedded-tests

kfaraz opened the pull request:
### Summary

- This patch migrates several indexing and compaction integration tests (several of which have been flaking out a lot recently) to the embedded-tests framework.
- Overall, this migration will save about __1 hr 30 mins__ of GitHub runner time.
- The MiddleManager version of the integration tests has currently not been ported as they are slower and currently not desirable in embedded tests (because they launch child processes).
- The MM-based embedded-tests can be easily enabled at any point in the future.

### Changes

- The new tests use SQL queries to verify results instead of native since the SQL syntax is much more concise and makes for more readable unit tests. We can add native queries later or in future tests that are migrated to this framework.
- Add `TaskBuilder` utility to create `Task` objects using fluent syntax
- Add JSON data resource files to `embedded-tests`

### Test migrations

|Old test|New test|
|--------|---------|
|`ITPerfectRollupParallelIndexTest`|`IndexParallelTaskTest`|
|`ITBestEffortRollupParallelIndexTest`|`IndexParallelTaskTest`, added as a new test parameter which uses dynamic partitioning|
|`ITAutoCompactionTest`|`AutoCompactionTest`|
|`ITAutoCompactionLockContentionTest`|`AutoCompactionUpgradeTest`|
|`ITAutoCompactionLockContentionTest`|`KafkaClusterMetricsTest`, method `test_ingestClusterMetrics_compactionSkipsLockedIntervals()`|
|`ITCompactionTaskTest`|`CompactionTaskTest`|
|`ITCompactionSparseColumnTest`|`CompactionSparseColumnTest`|
|`ITOverlordResourceTest`|Already being verified in `OverlordClientTest`|
|`ITOverlordResourceNotFoundTest`|Already being verified in `OverlordClientTest`|


### New nested tests

- `EmbeddedCentralizedSchemaPublishFailureTest` for the group `cds-task-schema-publish-disabled`
- `EmbeddedCentralizedSchemaMetadataQueryDisabledTest` for the group `cds-coordinator-metadata-query-disabled`

### Test run times

|Test|Actual test time|Total time including setup|
|----|-------------|-----------------------------|
|`IndexParallelTaskTest`(indexer)|38 s|__38 s__ |
|`AutoCompactionTest` (indexer)|1 min 20 s|__1 min 20 s__|
|`CompactionSparseColumnTest` (indexer)|12 s|__12 s__|
|Standard `ITPerfectRollupParallelIndexTest`|10 min|15 min|
|Standard `ITPerfectRollupParallelIndexTest`<br/>(Indexer, shuffle deep store test, only 1 config changed)|10 min|15 min|
|Standard `ITPerfectRollupParallelIndexTest`<br/>(MM, shuffle deep store test, only 1 config changed)|10 min|15 min|
|Standard `ITBestEffortRollupParallelIndexTest`|2 min| NA(setup includes other tests too) |
|Revised `ITBestEffortRollupParallelIndexTest`|4 min 10 s| " |
|Standard `ITAutoCompactionTest` (middle manager)|25 mins|"| 
|Standard `ITAutoCompactionTest` (indexer)|15 mins|"|
|Standard `ITCompactionSparseColumnTest` (indexer)|2 min 10 s|"|

---

This PR has:

- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

kfaraz commented on embedded-tests/src/test/resources/data/json/wikipedia_1.json:
```diff
@@ -0,0 +1,3 @@
+{"timestamp": "2013-08-31T01:02:33Z", "page": "Gypsy Danger", "language" : "en", "tags": ["t1", "t2"], "user" : "nuclear", "unpatrolled" : "true", "newPage" : "true", "robot": "false", "anonymous": "false", "namespace":"article", "continent":"North America", "country":"United States", "region":"Bay Area", "city":"San Francisco", "added": 57, "deleted": 200, "delta": -143}
```
Thanks for the suggestion! I will rename these datasets accordingly.

clintropolis commented on embedded-tests/src/test/java/org/apache/druid/testing/embedded/compact/EmbeddedAutoCompactionTest.java:
```diff
@@ -85,68 +90,177 @@
-  private static final String INDEX_TASK_WITHOUT_ROLLUP_FOR_PRESERVE_METRICS = "/indexer/wikipedia_index_no_rollup_preserve_metric.json";
+  private static final Logger LOG = new Logger(EmbeddedAutoCompactionTest.class);
+  private static final Supplier<TaskPayload> INDEX_TASK =
+      () -> TaskPayload
```
hm, if we're going to build these with code why not just use the real types to make stuff? making maps like this feels error prone, and we lose out on all the nice builders that already exist for some of the stuff

kfaraz replied:
Yeah, that's a fair point.

I think I started out with this so that the client side didn't need to add a bunch of modules to the ObjectMapper.
I remember something had failed due to the absence of some types.
However, since we are using the `HttpClient` and `ObjectMapper` from one of the servers itself, I can't recall why it had failed exactly.

Let me try using some of the existing builders.
Could you point me to the builder (if one exists) for a general `IndexTask` and `ParallelIndexSupervisorTask`?
If one doesn't exist, I will just try to convert the new builder in this PR to suit that purpose.

clintropolis replied:
ah i guess maybe there aren't any builders for the top level tasks, but much of the schema and config stuff at least has builders

kfaraz replied:
Hmm, let me write something up then.
It would make sense to have a clean builder that can be used in other UTs too.

I think this same builder can be repurposed to do that just by returning a concrete object rather than a map.

kfaraz commented on embedded-tests/src/test/java/org/apache/druid/testing/embedded/indexing/EmbeddedIndexParallelTaskTest.java:
```diff
@@ -0,0 +1,208 @@
+/**
+ * Runs tasks of "index_parallel" type using Indexers.
+ */
+public class EmbeddedIndexParallelTaskTest extends EmbeddedClusterTestBase
```
No, I generally like it mostly because it makes it easy to find and identify the tests and also to segregate them into a separate CI job when needed later (since they can be sprinkled across different modules).

clintropolis replied:
to clarify, none of this discussion is about changes to make in this PR btw, just thinking about stuff in general

kfaraz replied:
> wondering if we should still call these integration tests, since they really are basically, just using a bit different framework to run them than the other frameworks.

Yeah, I don't mind calling a test, say `IndexTaskIntegrationTest` or even any other name really, as long as we follow a single scheme across all modules. (I would just prefer not to retain the existing nomencluture of `IT*Test` as it always felt redundant 😛 ).

I personally liked the ring of `Embedded*Test` and it has the added advantage of denoting that the test uses an embedded cluster (along with the identification and segregation benefits that I mentioned above).
Since it uses an entire cluster, the fact that it is an integration test rather than a plain unit test is also implicit.
This name also felt appropriate since the primary module is called `embedded-tests`.

Let me know if you have any particular reservations against the current nomenclature or if you feel `*IntegrationTest` (or other similar names) would offer any specific cognitive advantage.

kfaraz replied:
As discussed offline, removed the `Embedded*` prefix to simplify the test names.

kfaraz commented:
Thanks for the feedback, @clintropolis !

I have made the following changes:
- Renamed the json data files to the format `tiny_wiki_x.json`
- Added `TaskBuilder` which creates `Task` objects and uses various existing builders. I have retained some utility methods for shorthand.

gianm commented on embedded-tests/src/test/java/org/apache/druid/testing/embedded/compact/AutoCompactionTest.java:
```diff
@@ -1989,6 +1884,7 @@ private void forceTriggerAutoCompaction(
       );
 
       // Wait for scheduler to pick up the compaction job
+      // TODO: make this latch-based
```
Yes, please don't commit todo comments.

kfaraz replied:
Sorry, must have missed this in the cleanup.

gianm commented on embedded-tests/src/test/java/org/apache/druid/testing/embedded/compact/CompactionTestBase.java:
```diff
@@ -0,0 +1,134 @@
+   *
+   * @return ID of the task.
+   */
+  protected String runTask(TaskBuilder<?, ?, ?> taskBuilder, String dataSource)
```
IMO, `protected` methods on a base class aren't the best way to do utility APIs. Other test cases might want some of these utility methods, but not to be "compaction tests". Also, some tests might want to extend multiple base classes, and this approach makes it impossible.

With MSQ tests we have `EmbeddedMSQApis`, something that collects together utility APIs without using a base class approach. Something similar might work here?

gianm commented on embedded-tests/src/test/java/org/apache/druid/testing/embedded/indexing/Resources.java:
```diff
@@ -0,0 +1,140 @@
+/**
+ * Constants and utility methods used in embedded cluster tests.
+ */
+public class Resources
```
Moving this to `embedded-tests` means that it's now tougher for extensions to write their own tests that pull in `Resources`. Can you split this up? Such as having some `Resources` in `services`, and `MoreResources` in `embedded-tests`.

kfaraz commented on embedded-tests/src/test/java/org/apache/druid/testing/embedded/compact/AutoCompactionTest.java:
```diff
@@ -85,68 +88,141 @@
+          .ofTypeIndexParallel()
+          .jsonInputFormat()
+          .inlineInputSourceWithData(Resources.InlineData.JSON_2_ROWS)
+          .isoTimestampColumn("timestamp")
```
I did have that initially. But realized that most tests are just using an ISO timestamp. So used this syntax sugar instead.

For a fully custom timestamp, there is also the option to use.
```java
.dataSchema(d -> d.withTimestamp(new TimestampSpec(...)))
```

Please let me know which one you prefer.

gianm replied:
I'm ok with either one.

kfaraz commented on embedded-tests/src/test/java/org/apache/druid/testing/embedded/compact/CompactionTestBase.java:
```diff
@@ -0,0 +1,134 @@
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+
+public abstract class CompactionTestBase extends EmbeddedClusterTestBase
```
Fair enough, I will move the common utility methods to some utility class similar to `EmbeddedMsqApis` as suggested by @gianm [here](https://github.com/apache/druid/pull/18207#discussion_r2237965907)

kfaraz commented on embedded-tests/src/test/java/org/apache/druid/testing/embedded/indexing/IndexTaskTest.java:
```diff
@@ -50,7 +50,7 @@
 /**
  * Simulation tests for batch {@link IndexTask} using inline datasources.
```
Thanks for spotting these!

kfaraz commented:
> One thing I was thinking about while reviewing is that i believe we are perhaps losing some minor coverage of auth stuff during this transition since I think the base ITs had basic-auth setup, though afaict not much in the way of roles and stuff in most tests (so i think was only authentication that would have been tested except for those that extend AbstractAuthConfigurationTest). I think that is probably fine though as long as we migrate the auth tests over to run on this framework, though they probably don't cast quite a wide of net in terms of APIs being called since those tests are more focused on authorization. I don't think I am suggesting we just bake basic auth into random tests or anything, and really its a bit of a negative of the old frameworks that you have to hunt across several files to determine what configuration is actually active for a given test, but maybe something we should watch out for as we move tests over.

Thanks for calling this out, @clintropolis !

I do plan to migrate the auth tests as well to the embedded framework.
For now, we will migrate the non-auth tests without basic auth enabled.
Once all tests are migrated, we can add test variants which have auth enabled to increase API coverage.

gianm commented:
> @gianm , @clintropolis , thanks for the feedback! I have updated the PR based on your suggestions. Please let me know if further changes are required.

Approved, since the only remaining comment (about `verifyScanResult`) is nonblocking.
